packages/react-router/modules/matchPath.js

```diff
@@ -26,15 +27,25 @@ const compilePath = (pattern, options) => {
 /**
  * Public API for matching a URL pathname to a path pattern.
  */
-const matchPath = (pathname, options = {}) => {
+const matchPath = (pathname, options = {}, parentMatch = null) => {
   if (typeof options === 'string')
     options = { path: options }
 
   const { exact = false, strict = false } = options
-  const path = options.path || options.from
+  let path = options.path != null ? options.path : options.from
+  const isRelative = !isAbsolute(path)
 
-  if (!path)
-    return { url: pathname, isExact: true, params: {} }
+  if (path == null)
+    return { url: parentMatch ? parentMatch.url : '/', isExact: true, params: {} }
+
+
+  warning(
+    !(isRelative && parentMatch == null),
+    'Relative paths will not be resolved when their parent match is null. You are most likely attempting to use a relative path inside of a parent <Route> that has a "children" prop, but that <Route>\'s path does not match the current location.'
+  )
```

Comment:
Added a warning. I don't think that this will be an issue too often, but a nice message for people that run into this problem could be useful.